### PR TITLE
feat: add azure pipelines built-in tasks version data

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes

Adds version information for Azure Pipelines built-in tasks.

There is no public API that exposes this information. I checked with the team that maintains these tasks, and parsing the git history was the recommended way to extract version information. I don't believe doing this during a Renovate run is feasible, hence adding it as static information in the repository.

The script currently catches ~3~ 2 exceptions throughout the `azure-pipelines-tasks` repository's history:
- `Failed to parse Tasks/CondaEnvironmentV0/task.json at 7ff8f87757903515cdcf521a76f92626bd06ab2a`
	- `CondaEnvironmentV0/task.json` was moved to `CondaEnvironmentV1/task.json`
- ~`Failed to parse Tasks/DownloadPipelineArtifactV2/task.json at d0c0a1cce48931d8ba009ef3db51651c47b0c924`~
	- ~Someone added a comment to `DownloadPipelineArtifactV2/task.json`~
- `Failed to parse Tasks/IISWebAppDeployment/task.json at 05114960d2286a029cf1cbd99380df3a8332e9be`
	- The file was deleted (and later re-added)

If there are any git wizards who know how to exclude renames and deletions from `git rev-list` please let me know.

## Context

- Part of #15818
- Required by #16904

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  for (const file of files) {
+    const revs = (await git.raw(['rev-list', 'master', '--', file])).split(
+      '\n'
+    );
+    shell.echo(`Parsing ${file}`);
+    for (const rev of revs) {
+      try {
+        const content = await git.show([`${rev}:${file}`]);
+        /** @type {{name: string, version: {Major: number, Minor: number, Patch: number}}} */
+        const parsedContent = JSON.parse(content);
```
Comment (viceice) [inline tools/azure-pipelines-tasks-json-generate.mjs:42]:
```suggestion
        const parsedContent = JSON5.parse(content);
```
to fix the comments issue

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  const files = (await glob(path.join(localPath, '**/task.json'))).map((file) =>
+    file.replace(`${localPath}/`, '')
+  );
+
+  /** @type {Record<string, Set<string>>} */
+  const tasks = {};
+
+  for (const file of files) {
+    const revs = (await git.raw(['rev-list', 'master', '--', file])).split(
```
Comment (viceice) [inline tools/azure-pipelines-tasks-json-generate.mjs:34]: we should detect main branch if it's changed in future

Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:34]: `git rev-list HEAD` gives equal results and sidesteps the issue. I assume `git clone` will pull the default branch.

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+const glob = promisify(g);
+
+const localPath = path.join(os.tmpdir(), 'azure-pipelines-tasks');
+
+await (async () => {
+  await fs.ensureDir(localPath);
+  const git = Git(localPath);
+
+  if (await git.checkIsRepo()) {
+    await git.pull();
```
Comment (rarkins) [inline tools/azure-pipelines-tasks-json-generate.mjs:19]: There's some chance that a local git repo could get into an unpullable state, e.g. if someone force pushed to the origin? But if this doesn't affect our automated systems then it's no big deal and can be resolved by the user through manual deletion

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  } else {
+    await git.clone(
+      'https://github.com/microsoft/azure-pipelines-tasks.git',
+      '.'
+    );
```
Comment (rarkins) [inline tools/azure-pipelines-tasks-json-generate.mjs:24]: Are you sure that a full clone is necessary and no other solution such as blobless cloning suffices?

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  const files = (await glob(path.join(localPath, '**/task.json'))).map((file) =>
+    file.replace(`${localPath}/`, '')
+  );
+
+  /** @type {Record<string, Set<string>>} */
+  const tasks = {};
+
+  for (const file of files) {
+    const revs = (await git.raw(['rev-list', 'HEAD', '--', file])).split('\n');
```
Comment (rarkins) [inline tools/azure-pipelines-tasks-json-generate.mjs:35]: Can this be documented? I'm not sure what the purpose is. Is it just the content of the file locally, in which case why the need to use git?

**Review by rarkins — CHANGES_REQUESTED** [review]
I think this algorithm you created needs some documentation to explain the "why", in case someone else needs to maintain it :)

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 4 above)
Comment (rarkins) [reply tools/azure-pipelines-tasks-json-generate.mjs:24]: It's pretty big:

```
❯ git clone https://github.com/microsoft/azure-pipelines-tasks.git
Cloning into 'azure-pipelines-tasks'...
remote: Enumerating objects: 270480, done.
remote: Total 270480 (delta 0), reused 0 (delta 0), pack-reused 270480
Receiving objects: 100% (270480/270480), 482.08 MiB | 6.81 MiB/s, done.
Resolving deltas: 100% (160273/160273), done.
```

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 3 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:19]: I think it's unlikely that someone will do that, given that it's a protected branch that requires PRs. Like you said, not an issue for automated systems, and can easily be resolved locally.

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 4 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:24]: I experimented with a couple of optimized clones. Here's the current full clone from scratch

```
$ time node tools/azure-pipelines-tasks-json-generate.mjs
...
________________________________________________________
Executed in  122.07 secs    fish           external
   usr time   94.82 secs  264.00 micros   94.82 secs
   sys time   48.89 secs  652.00 micros   48.89 secs
```

I experimented with `--filter=blob:none`, `--filter=tree:0`, and `sparse-checkout`. `--filter=blob:none` and `--filter=tree:0` result in quicker initial clones, but massively slow down the overall script due to doing thousands of tiny requests on demand. I don't have timing data for these as I killed them after 2 minutes.

Here's the `sparse-checkout`:

```
$ time node tools/azure-pipelines-tasks-json-generate.mjs
________________________________________________________
Executed in  116.29 secs    fish           external
   usr time   98.13 secs  182.00 micros   98.13 secs
   sys time   45.41 secs  456.00 micros   45.41 secs
```

IMO it's not worth the additional complexity of sparse clones for what is within the margin of error.

Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:24]: Unfortunately we do need the entire history to find all the historical versions.

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 5 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:35]: `git rev-list HEAD -- ${file}` gets a list of every commit in which the specific `task.json` was modified. `git show ${rev}:${file}` gets the content of that file at the specific commit. To summarize this script, it gets every revision of every `task.json` and extracts the version from it. Not every revision updates the version, and that's why I use a set to de-duplicate versions.

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      } catch (e) {
+        shell.echo(`Failed to parse ${file} at ${rev}`);
+        shell.echo(e.toString());
+      }
+    }
+  }
+
+  const data = JSON.stringify(
+    tasks,
+    (_, value) => (value instanceof Set ? [...value] : value),
```
Comment (viceice) [inline tools/azure-pipelines-tasks-json-generate.mjs:54]: 🤔 isn't the serializer smart enough to handle set?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 6 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:54]: Unfortunately not. It's a [known issue](https://stackoverflow.com/questions/31190885/json-stringify-a-set)

```
let a = new Set([1, 2, 3])
undefined

JSON.stringify(a)
"{}"

JSON.stringify(a, (_, value) => (value instanceof Set ? [...value] : value))
"[1,2,3]"
```

Comment (viceice) [reply tools/azure-pipelines-tasks-json-generate.mjs:54]: 😕

**Review by viceice — DISMISSED** [review]

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 5 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:44]: I've added comments explaining how the script works, in case anyone ever needs to update it.

**Review by rarkins — COMMENTED** [review]
Considering that it seems quite slow, how/when will this command need to be called?

**Comment by JamieMagee** [comment]
I think once a week should cover it. Shall I create a follow-up task to write a weekly scheduled GitHub Actions workflow?

**Comment by rarkins** [comment]
Yeah, that would be good. I just wanted to make sure it doesn't run as part of `yarn test` or any time like that. If it's only manually invoked then that's fine by me

**Comment by JamieMagee** [comment]
This PR adds the script, but doesn't run it anywhere. It requires a manual run, same as the `update:distro-info` script we have to update Debian and Ubuntu version information.

**Review by rarkins — APPROVED** [review]

File: tools/azure-pipelines-tasks-json-generate.mjs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    const revs = (await git.raw(['rev-list', 'HEAD', '--', file])).split('\n');
+    shell.echo(`Parsing ${file}`);
+    for (const rev of revs) {
+      try {
+        // Get the content of the file at the commit
+        const content = await git.show([`${rev}:${file}`]);
+        /** @type {{name: string, version: {Major: number, Minor: number, Patch: number}}} */
+        const parsedContent = JSON5.parse(content);
+        const version = `${parsedContent.version.Major}.${parsedContent.version.Minor}.${parsedContent.version.Patch}`;
+        tasks[parsedContent.name.toLowerCase()] =
```
Comment (viceice) [inline tools/azure-pipelines-tasks-json-generate.mjs:53]: i think we should use upper case, as there are chars which doesn't have a lower case variant

**Comment by viceice** [comment]
> I think once a week should cover it. Shall I create a follow-up task to write a weekly scheduled GitHub Actions workflow?

sounds good, like that. how should it notify us of changes? ideally with an update PR. 🤗

File: tools/azure-pipelines-tasks-json-generate.mjs (hunk 7 above)
Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:53]: In general I agree, but task names are restricted to `^[A-Za-z0-9_]+$`. So in this specific instance I think we are okay.

Comment (JamieMagee) [reply tools/azure-pipelines-tasks-json-generate.mjs:53]: In general I agree, but task names are restricted to `^[A-Za-z0-9_]+$`. So in this specific instance I think we are okay.

**Pull request merged**